plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+        }
+        return referencePackageFunction("kotlinx.atomicfu", "$ATOMICFU_RUNTIME_FUNCTION_PREDICATE$functionName") {
+            val typeArg = it.owner.getGetterReturnType()
+            !(typeArg as IrType).isPrimitiveType() || typeArg == type
+        }
+    }
+
+    private fun IrFunction.getGetterReturnType() = (valueParameters[valueParameters.lastIndex - 1].type as IrSimpleType).arguments.first().typeOrNull
+
+    private fun IrCall.isAtomicFactory(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_CONSTRUCTOR
+    }
+
+    private fun IrCall.isAtomicArrayFactory(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_ARRAY_FACTORY_FUNCTION
+    }
+
+    private fun IrConstructorCall.isAtomicArrayConstructor() = (type as IrSimpleType).isAtomicArrayType()
+
+    private fun IrSymbol.isKotlinxAtomicfuPackage() =
+        this.isPublicApi && signature?.packageFqName()?.asString() == AFU_PKG.prettyStr()
+
+    private fun IrType.isAtomicValueType() = belongsTo(ATOMICFU_VALUE_TYPE)
+    private fun IrType.isAtomicArrayType() = belongsTo(ATOMIC_ARRAY_TYPE)
+    private fun IrType.isReentrantLockType() = belongsTo("$AFU_PKG/$LOCKS", REENTRANT_LOCK_TYPE)
+
+    private fun IrType.belongsTo(typeName: String) = belongsTo(AFU_PKG, typeName)
+
+    private fun IrType.belongsTo(packageName: String, typeName: String): Boolean {
+        if (this !is IrSimpleType || !(classifier.isPublicApi && classifier is IrClassSymbol)) return false
+        val signature = classifier.signature?.asPublic() ?: return false
+        val pckg = signature.packageFqName().asString()
+        val type = signature.declarationFqName
+        return pckg == packageName.prettyStr() && type.matches(typeName.toRegex())
+    }
+
+    private fun IrCall.getAtomicFunctionName(): String {
+        val signature = symbol.signature!!
+        val classFqName = if (signature is IdSignature.AccessorSignature) {
+            signature.accessorSignature.declarationFqName
+        } else (signature.asPublic()!!).declarationFqName
+        val pattern = "$ATOMICFU_VALUE_TYPE\\.(.*)".toRegex()
+        return pattern.findAll(classFqName).firstOrNull()?.let { it.groupValues[2] } ?: classFqName
+    }
+
+    private fun IrType.atomicToValueType(): IrType {
+        require(isAtomicValueType())
+        // todo check correctness of this way to get type name
+        val classId = ((this as IrSimpleType).classifier.signature!!.asPublic())!!.declarationFqName
+        if (classId == "AtomicRef") {
+            return arguments.first().typeOrNull ?: error("$AFU_PKG/AtomicRef type parameter is not IrTypeProjection")
+        }
+        return AFU_CLASSES[classId] ?: error("IrType ${this.getClass()} does not match any of atomicfu types")
+    }
+
+    private fun buildConstNull() = IrConstImpl.constNull(UNDEFINED_OFFSET, UNDEFINED_OFFSET, context.irBuiltIns.anyType)
```

Comment:
Since `null` is obviously nullable `anyNType` has to be used instead of `anyType`